### PR TITLE
[SYCLomatic] Support migrate source code string without file.

### PR DESCRIPTION
**Pull request by tangjj11**
Signed-off-by: Tang, Jiajun jiajun.tang@intel.com

### CODE REVIEW

File: clang/lib/DPCT/DPCT.cpp [hunk 1]
```diff
@@ -733,19 +733,24 @@ int runDPCT(int argc, const char **argv) {
 
   std::vector<std::string> SourcePathList;
   if (QueryAPIMapping.getNumOccurrences()) {
+    // Set a virtual file for --query-api-mapping.
+    llvm::SmallVector<char> SysTempDir;
```
Comment (yihwang-nv) [inline clang/lib/DPCT/DPCT.cpp]: Let's use `llvm::SmallString`

**Review by tomflinda — APPROVED** [review]
LGTM

File: clang/lib/DPCT/DPCT.cpp [hunk 2]
```diff
@@ -733,19 +733,24 @@ int runDPCT(int argc, const char **argv) {
...
+    llvm::sys::path::system_temp_directory(true, SysTempDir);
+    SourcePathList.emplace_back(
+        std::string(SysTempDir.data(), SysTempDir.size()) + "/temp.cu");
```
Comment (yihwang-nv) [inline clang/lib/DPCT/DPCT.cpp]: Let's use `llvm::SmallString::str()`

File: clang/lib/DPCT/DPCT.cpp [hunk 3]
```diff
@@ -1098,7 +1099,8 @@ int runDPCT(int argc, const char **argv) {
          I.MoveToNextPiece()) {
       if (Flag) {
         if (auto It = I.piece().find(EndStr); It != StringRef::npos) {
-          llvm::outs() << I.piece().substr(0, It);
+          auto TempStr = I.piece().substr(0, It);
+          llvm::outs() << TempStr.substr(0, TempStr.find_last_of("\n") + 1);
```
Comment (yihwang-nv) [inline clang/lib/DPCT/DPCT.cpp:1103]: Is this line used to debug?

File: clang/lib/DPCT/MigrationAction.cpp [hunk 4]
```diff
@@ -126,11 +126,12 @@ void DpctFrontEndAction::EndSourceFileAction() {
   getCompilerInstance().getASTContext().getParentMapContext().clear();
 }
 
-DpctToolAction::DpctToolAction(llvm::raw_ostream &DS, ReplTy &Replacements,
-                               const std::string &RuleNames,
-                               std::vector<PassKind> Passes)
+DpctToolAction::DpctToolAction(
+    llvm::raw_ostream &DS, ReplTy &Replacements, const std::string &RuleNames,
+    std::vector<PassKind> Passes,
+    const llvm::IntrusiveRefCntPtr<llvm::vfs::FileSystem> FS)
```
Comment (yihwang-nv) [inline clang/lib/DPCT/MigrationAction.cpp]: Please remove `const` qualifier

File: clang/lib/DPCT/MigrationAction.h [hunk 5]
```diff
@@ -56,7 +56,8 @@ class DpctFrontEndAction : public ASTFrontendAction {
 class DpctToolAction : public tooling::ToolAction {
 public:
   DpctToolAction(llvm::raw_ostream &DS, ReplTy &Replacements,
-                 const std::string &RuleNames, std::vector<PassKind> Passes);
+                 const std::string &RuleNames, std::vector<PassKind> Passes,
+                 const llvm::IntrusiveRefCntPtr<llvm::vfs::FileSystem> FS);
```
Comment (yihwang-nv) [inline clang/lib/DPCT/MigrationAction.h]: likewise

File: clang/lib/DPCT/MigrationAction.h [hunk 6]
```diff
@@ -56,7 +56,8 @@ class DpctFrontEndAction : public ASTFrontendAction {
 class DpctToolAction : public tooling::ToolAction {
 public:
   DpctToolAction(llvm::raw_ostream &DS, ReplTy &Replacements,
-                 const std::string &RuleNames, std::vector<PassKind> Passes);
+                 const std::string &RuleNames, std::vector<PassKind> Passes,
```
Comment (yihwang-nv) [inline clang/lib/DPCT/MigrationAction.h:59]: I think we can use `StringRef`

File: clang/lib/DPCT/DPCT.cpp (hunk 1 above)
Comment (tangjj11) [reply clang/lib/DPCT/DPCT.cpp]: fixed

File: clang/lib/DPCT/DPCT.cpp (hunk 2 above)
Comment (tangjj11) [reply clang/lib/DPCT/DPCT.cpp]: fixed

File: clang/lib/DPCT/DPCT.cpp (hunk 3 above)
Comment (tangjj11) [reply clang/lib/DPCT/DPCT.cpp:1103]: no

File: clang/lib/DPCT/MigrationAction.cpp (hunk 4 above)
Comment (tangjj11) [reply clang/lib/DPCT/MigrationAction.cpp]: fixed

File: clang/lib/DPCT/MigrationAction.h (hunk 5 above)
Comment (tangjj11) [reply clang/lib/DPCT/MigrationAction.h]: fixed

File: clang/lib/DPCT/MigrationAction.h (hunk 6 above)
Comment (tangjj11) [reply clang/lib/DPCT/MigrationAction.h:59]: not changed.

File: clang/lib/DPCT/DPCT.cpp [hunk 7]
```diff
@@ -733,19 +733,23 @@ int runDPCT(int argc, const char **argv) {
 
   std::vector<std::string> SourcePathList;
   if (QueryAPIMapping.getNumOccurrences()) {
+    // Set a virtual file for --query-api-mapping.
+    llvm::SmallString<32> SysTempDir;
+    llvm::sys::path::system_temp_directory(true, SysTempDir);
```
Comment (yihwang-nv) [inline clang/lib/DPCT/DPCT.cpp]:
```suggestion
    llvm::sys::path::system_temp_directory(/*ErasedOnReboot=*/true, SysTempDir);
```

File: clang/lib/DPCT/DPCT.cpp [hunk 8]
```diff
@@ -733,19 +733,23 @@ int runDPCT(int argc, const char **argv) {
...
+    SourcePathList.emplace_back(SysTempDir.str().str() + "/temp.cu");
```
Comment (yihwang-nv) [inline clang/lib/DPCT/DPCT.cpp]:
```suggestion
    SourcePathList.emplace_back(llvm::Twine(SysTempDir).concat("/temp.cu").str());
```

**Review by yihwang-nv — APPROVED** [review]
LGTM

File: clang/lib/DPCT/DPCT.cpp (hunk 7 above)
Comment (tangjj11) [reply clang/lib/DPCT/DPCT.cpp]: fixed

File: clang/lib/DPCT/DPCT.cpp (hunk 8 above)
Comment (tangjj11) [reply clang/lib/DPCT/DPCT.cpp]: fixed

**Review by ziranzha — APPROVED** [review]

**Pull request merged**